fibsem/milling/patterning/patterns2.py

```diff
@@ -12,69 +12,83 @@
     FibsemBitmapSettings,
     FibsemCircleSettings,
     FibsemLineSettings,
-    FibsemPatternSettings,
     FibsemRectangleSettings,
     Point,
+    TFibsemPatternSettings,
 )
 
+TPattern = TypeVar("TPattern", bound="BasePattern")
+
 # TODO: define the configuration for each key,
 # e.g. 
 # "width": {"type": "float", "min": 0, "max": 1000, "default": 100, "description": "Width of the rectangle"}
 # "cross_section": {"type": "str", "options": [cs.name for cs in CrossSectionPattern], "default": "Rectangle", "description": "Cross section of the milling pattern"}
 
-DEFAULT_POINT_DDICT = {"x": 0.0, "y": 0.0}
-
 ####### Combo Patterns
 
-CORE_PATTERN_ATTRIBUTES = ["name", "point", "shapes"]
-
 @dataclass
-class BasePattern(ABC):
-    # name: str = "BasePattern"
-    # point: Point = field(default_factory=Point) 
-    # shapes: List[FibsemPatternSettings] = None
-    # TODO: investigate TypeError: non-default argument 'width' follows default argument when uncommenting the above lines
+class BasePattern(ABC, Generic[TFibsemPatternSettings]):
+    name: ClassVar[str] = field(init=False)
+    point: Point = field(default_factory=Point)
+    shapes: Optional[List[TFibsemPatternSettings]] = field(default=None, init=False)
 
-    @abstractmethod
-    def define(self) -> List[FibsemPatternSettings]:
-        pass
+    _advanced_attributes: ClassVar[Tuple[str, ...]] = ()
 
     @abstractmethod
-    def to_dict(self):
+    def define(self) -> List[TFibsemPatternSettings]:
         pass
 
-    @classmethod
-    @abstractmethod
-    def from_dict(cls, ddict: dict) -> "BasePattern":
-        pass
+    def to_dict(self) -> Dict[str, Any]:
+        ddict = asdict(self)
+        # Handle any special cases
+        if "cross_section" in ddict:
+            ddict["cross_section"] = ddict["cross_section"].name
+        ddict["name"] = self.name
+        del ddict["shapes"]
+        return ddict
 
+    @classmethod
+    def from_dict(cls: Type[TPattern], ddict: Dict[str, Any]) -> TPattern:
+        kwargs = {}
+        for f in fields(cls):
+            if f.name in ddict:
+                kwargs[f.name] = ddict[f.name]
+
+        # Construct objects
+        point = kwargs.pop("point", None)
+        if point is not None:
+            kwargs["point"] = Point.from_dict(point)
+
+        cross_section = kwargs.pop("cross_section", None)
+        if cross_section is not None:
+            kwargs["cross_section"] = CrossSectionPattern[cross_section]
+
+        return cls(**kwargs)
+                
     @property
-    def required_attributes(self) -> Tuple[str]:
-        return [field.name for field in fields(self) if field.name not in CORE_PATTERN_ATTRIBUTES]
+    def required_attributes(self) -> Tuple[str, ...]:
+        return tuple(f.name for f in fields(self) if f.name not in fields(BasePattern))
```

Comment:
Comparison condition is incorrect
```suggestion
        return tuple(f.name for f in fields(self) if f not in fields(BasePattern))
```

Reply:
Good catch - I've made the change now